### PR TITLE
BlockRepository and utility tests, small fixes

### PR DESCRIPTION
**Pull request by Neurosploit**
- Several new classes fully tested.
- Moved BlockNotificationFeatureTest to the Notifications folder.
- DatetimeProvider interface is now registered with it's interface in the services.
- BlockRepository Initialize has been moved outside the constructor. Updated the registration.
- updated the contribution guide links.

Please carefully review the changes regarding datetimeprovider and blockrepository initialize. All tests pass on my machine and it runs without any issues.

### CODE REVIEW

File: Stratis.Bitcoin.Tests/Utilities/AsyncDictionaryTest.cs [hunk 1]
```diff
@@ -0,0 +1,213 @@
+﻿using Moq;
```
Comment (dangershony) [inline Stratis.Bitcoin.Tests/Utilities/AsyncDictionaryTest.cs]: Great, its not use yet but with the AsyncDictionary we can remove locks in ConcurrentDictionary

File: Stratis.Bitcoin/BlockStore/BlockRepository.cs [hunk 2]
```diff
@@ -36,12 +44,13 @@ public BlockRepository(Network network, string folder)
 
 			this.session = new DBreezeSingleThreadSession("DBreeze BlockRepository", folder);
 			this.network = network;
-			Initialize(network.GetGenesis()).GetAwaiter().GetResult(); // hmm...
 		}
 
-		private Task Initialize(Block genesis)
```
Comment (dangershony) [inline Stratis.Bitcoin/BlockStore/BlockRepository.cs]: Next commit do the same fro DbCoinView (use the network from the constructor)

File: Stratis.Bitcoin/BlockStore/BlockStoreFeature.cs [hunk 3]
```diff
@@ -57,13 +58,16 @@ public static class MBlockStoreBuilderExtension
 	{
 		public static IFullNodeBuilder UseBlockStore(this IFullNodeBuilder fullNodeBuilder)
 		{
-			fullNodeBuilder.ConfigureFeature(features =>
+            var blockRepository = new BlockRepository(fullNodeBuilder.Network, new DataFolder(fullNodeBuilder.NodeSettings));
```
Comment (dangershony) [inline Stratis.Bitcoin/BlockStore/BlockStoreFeature.cs]: BlockRepository can to be injected, do the same as DbCoinView take DataFolder in an additional constructor.
DataFolder is already injected to the DI

File: Stratis.Bitcoin/Notifications/BlockNotification.cs [hunk 4]
```diff
@@ -37,9 +38,14 @@ public BlockNotification(ConcurrentChain chain, ILookaheadBlockPuller puller, Si
         /// <param name="startHash">The hash of the block from which to start notifying</param>
         /// <param name="cancellationToken">A cancellation token</param>
         public virtual void Notify(uint256 startHash, CancellationToken cancellationToken)
-        {            
+        {
+            ChainedBlock startBlock = this.Chain.GetBlock(startHash);
+
```
Comment (dangershony) [inline Stratis.Bitcoin/Notifications/BlockNotification.cs]: This code is changing slightly

Comment (bokobza) [inline Stratis.Bitcoin/Notifications/BlockNotification.cs]: Indeed but I'm happy for this to go in and to add my changes afterwards.

**Comment by bokobza** [comment]
Great PR and I have nothing to add to @dangershony's comments.

File: Stratis.Bitcoin/BlockStore/BlockStoreFeature.cs (hunk 3 above)
Comment (Neurosploit) [inline Stratis.Bitcoin/BlockStore/BlockStoreFeature.cs]: Yes I know that the datafolder and network are probably already registered services but I can't resolve them in the ConfigureFeature without the IServiceProvider. That would be even nicer.

Comment (dangershony) [inline Stratis.Bitcoin/BlockStore/BlockStoreFeature.cs]: I realize BlockRepository is already injected so really just call blockRepository.Initialize() from the feature strat

**Comment by dangershony** [comment]
It looks good to me @bokobza can you double check and push.

**Comment by bokobza** [comment]
All good with me.

**Pull request merged**